File: 028.Divide_Two_Integers/AC_binary_nlogn.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_binary_nlogn.cpp
*  Create Date: 2015-01-16 09:42:12
*  Descripton:  Binary. Just use the bit.
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

typedef long long ll;

class Solution {
public:
    int divide(int dividend, int divisor) {
		ll a = dividend >= 0 ? dividend : -(ll)dividend;
		ll b = divisor >= 0 ? divisor : -(ll)divisor;
		ll result = 0, c = 0;
		bool sign = (dividend > 0 && divisor < 0) ||
			(dividend < 0 && divisor > 0);

		while (a >= b) {
			c = b;
			for (int i = 0; a >= c; i++, c <<= 1) {
				a -= c;
				result += (1<<i);
			}
		}
		if (result > INT_MAX)
			result = INT_MAX;
		return sign ? -result : result;
    }
};

int main() {
	int a, b;
	Solution s;
	while (cin >> a >> b) {
		cout << s.divide(a, b) << endl;
	}
	return 0;
}

File: solutions/029.Divide_Two_Integers/AC_binary_nlogn.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_binary_nlogn.py
# Create Date: 2015-03-06 15:28:28
# Usage:       AC_binary_nlogn.py 
# Descripton:  


class Solution:
    # @return an integer
    def divide(self, dividend, divisor):
        sign = (dividend < 0 and divisor > 0) or (dividend > 0 and divisor < 0)
        a, b = abs(dividend), abs(divisor)
        ret, c = 0, 0

        while a >= b:
            c = b
            i = 0
            while a >= c:
                a -= c
                ret += (1<<i)
                i += 1
                c <<= 1

        if sign:
            ret = -ret
        return min(max(-2147483648, ret), 2147483647)

s = Solution()
print s.divide(1, 1)
